packetbeat/protos/amqp/amqp.go

```diff
@@ -0,0 +1,883 @@
+			EXCHANGEDELETEOK:  okMethod,
+			EXCHANGEBIND:      exchangeBindMethod,
+			EXCHANGEBINDOK:    okMethod,
+			EXCHANGEUNBIND:    exchangeUnbindMethod,
+			EXCHANGEUNBINDOK:  okMethod,
+		},
+		QUEUE: map[uint16]AmqpMethod{
+			QUEUEDECLARE:   queueDeclareMethod,
+			QUEUEDECLAREOK: queueDeclareOkMethod,
+			QUEUEBIND:      queueBindMethod,
+			QUEUEBINDOK:    okMethod,
+			QUEUEUNBIND:    queueUnbindMethod,
+			QUEUEUNBINDOK:  okMethod,
+			QUEUEPURGE:     queuePurgeMethod,
+			QUEUEPURGEOK:   queuePurgeOkMethod,
+			QUEUEDELETE:    queueDeleteMethod,
+			QUEUEDELETEOK:  queueDeleteOkMethod,
+		},
+		BASIC: map[uint16]AmqpMethod{
+			BASICCONSUME:   basicConsumeMethod,
+			BASICCONSUMEOK: basicConsumeOkMethod,
+			BASICCANCEL:    basicCancelMethod,
+			BASICCANCELOK:  basicCancelOkMethod,
+			BASICPUBLISH:   basicPublishMethod,
+			BASICRETURN:    basicReturnMethod,
+			BASICDELIVER:   basicDeliverMethod,
+			BASICGET:       basicGetMethod,
+			BASICGETOK:     basicGetOkMethod,
+			BASICGETEMPTY:  basicGetEmptyMethod,
+			BASICACK:       basicAckMethod,
+			BASICREJECT:    basicRejectMethod,
+			BASICRECOVER:   basicRecoverMethod,
+			BASICRECOVEROK: okMethod,
+			BASICNACK:      basicNackMethod,
+		},
+		TX: map[uint16]AmqpMethod{
+			TXSELECT:     txSelectMethod,
+			TXSELECTOK:   okMethod,
+			TXCOMMIT:     txCommitMethod,
+			TXCOMMITOK:   okMethod,
+			TXROLLBACK:   txRollbackMethod,
+			TXROLLBACKOK: okMethod,
+		},
+	}
+}
+
+func (amqp *Amqp) Init(test_mode bool, results publish.Transactions) error {
+	amqp.InitDefaults()
+	if !test_mode {
+		err := amqp.setFromConfig(config.ConfigSingleton.Protocols.Amqp)
+		if err != nil {
+			return err
+		}
+	}
+	if amqp.Hide_connection_information == false {
+		amqp.addConnectionMethods()
+	}
+	amqp.transactions = common.NewCache(
+		amqp.transactionTimeout,
+		protos.DefaultTransactionHashSize)
+	amqp.transactions.StartJanitor(amqp.transactionTimeout)
+	amqp.results = results
+	return nil
+}
+
+func (amqp *Amqp) addConnectionMethods() {
+	amqp.MethodMap[CONNECTION][CONNECTIONSTART] = connectionStartMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONSTARTOK] = connectionStartOkMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONTUNE] = connectionTuneMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONTUNEOK] = connectionTuneOkMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONOPEN] = connectionOpenMethod
+	amqp.MethodMap[CONNECTION][CONNECTIONOPENOK] = okMethod
+	amqp.MethodMap[CHANNEL][CHANNELOPEN] = channelOpenMethod
+	amqp.MethodMap[CHANNEL][CHANNELOPENOK] = okMethod
+	amqp.MethodMap[CHANNEL][CHANNELFLOW] = channelFlowMethod
+	amqp.MethodMap[CHANNEL][CHANNELFLOWOK] = channelFlowOkMethod
+	amqp.MethodMap[BASIC][BASICQOS] = basicQosMethod
+	amqp.MethodMap[BASIC][BASICQOSOK] = okMethod
+}
+
+func (amqp *Amqp) ConnectionTimeout() time.Duration {
+	return amqp.transactionTimeout
+}
+
+func (amqp *Amqp) Parse(pkt *protos.Packet, tcptuple *common.TcpTuple,
+	dir uint8, private protos.ProtocolData) protos.ProtocolData {
```

Comment:
common pattern on stream initialization, already changed in some modules to use pointers and split stream init from parsing. E.g. https://github.com/elastic/beats/blob/master/packetbeat/protos/redis/redis.go#L99